### PR TITLE
adding support for array shape () passed as an index argument in advanced indexing

### PR DESCRIPTION
**Pull request by ipdemes**
This PR making following work:
```
    arr[offsets[3]]=-1
```

### CODE REVIEW

File: tests/integration/test_set_item.py [hunk 1]
```diff
@@ -40,9 +40,9 @@ def test_basic():
 def test_scalar_ndarray_as_index(arr):
     offsets = num.arange(5)  # [0, 1, 2, 3, 4]
     offset = offsets[3]  # 3
-    # arr[offset] = -1  # TODO: doesn't work when arr is a num.ndarray
+    arr[offset] = -1
     arr[offset - 2 : offset] = [-1, -1]
-    assert np.array_equal(arr, [4, -1, -1, 1, 0])
+    assert np.array_equal(arr, [4, -1, -1, -1, 0])
```
Comment (manopapad) [inline tests/integration/test_set_item.py:47]: Please also uncomment the corresponding get_item test in `test_get_item.py`

**Comment by manopapad** [comment]
This has different behavior from NumPy on `get_item`. This code prints out `[1]`:

```
arr = 4 - cn.arange(5)
offsets = cn.arange(5)
offset = offsets[3]
print(arr[offset])
```

while NumPy returns `1`:

```
arr = 4 - np.arange(5)
offsets = np.arange(5)
offset = offsets[3]
print(arr[offset])
```

I naively just tried removing the changes on deferred.py (and keeping just the C++ changes), but that triggers https://github.com/nv-legate/legate.core/issues/314.

**Comment by ipdemes** [comment]
@manopapad : Thank you for finding this! I will wait for Wonchan to update on #314

**Comment by ipdemes** [comment]
@manopapad I have added fixes for few other cases when either lhs or rhs are futures

File: tests/integration/test_get_item.py [hunk 2]
```diff
@@ -36,7 +36,7 @@ def test_basic():
 def test_scalar_ndarray_as_index(arr):
     offsets = num.arange(5)  # [0, 1, 2, 3, 4]
     offset = offsets[3]  # 3
-    # assert arr[offset] == 1  # TODO: doesn't work when arr is a num.ndarray
+    assert arr[offset] == 1
```
Comment (manopapad) [inline tests/integration/test_get_item.py:39]: Actually, can we do this check as follows?
```
assert np.array_equal(arr[offset], 1)
```
This way we also check that a scalar is returned, whereas the original `assert arr[offset] == 1` will succeed if `arr[offset]` returns an array of shape `(1,)`.

Comment (ipdemes) [reply tests/integration/test_get_item.py:39]: done

**Comment by ipdemes** [comment]
@manopapad @magnatelee : do you think we can still merge this to 22.07 since if fixes some corner cases of advanced indexing? Or should I change this PR to be merged to 22.10?

File: cunumeric/deferred.py [hunk 3]
```diff
@@ -463,7 +463,8 @@ def _zip_indices(
         task.add_scalar_arg(self.shape, (ty.int64,))
         for a in arrays:
             task.add_input(a)
-            task.add_alignment(output_arr.base, a)
+            if a.shape != ():
+                task.add_alignment(output_arr.base, a)
```
Comment (manopapad) [inline cunumeric/deferred.py:467]: I don't think this is necessary. Has this caused issues?

I am a little hesitant to skip alignment, even for seemingly trivial cases, as I don't know how the auto-partitioner will handle an under-constrained case like this. @magnatelee any comments?

File: cunumeric/deferred.py [hunk 4]
```diff
@@ -737,7 +738,11 @@ def get_item(self, key: Any) -> NumPyThunk:
                 index_array,
                 self,
             ) = self._create_indexing_array(key)
+
+            if rhs.base.kind == Future:
+                rhs = self._convert_future_to_store(rhs)
```
Comment (manopapad) [inline cunumeric/deferred.py:743]: You can move this inside the `if copy_needed` branch.

File: cunumeric/deferred.py (hunk 3 above)
Comment (ipdemes) [reply cunumeric/deferred.py:467]: Actually, we don't need `if a.shape != ():` anymore. I will remove it in a minute

**Comment by manopapad** [comment]
I added some further tests involving set/getitem on scalar arrays. It looks like there are issues when the index or value array is  0d and backed by a region-field:

```
arr = cn.full((5,), 42)
idx = cn.zeros((3,), dtype=np.int64)[2:3].reshape(())
arr[idx]
```

This example asserts inside of Legion:

```
[0 - 70000c362000]    2.048913 {5}{runtime}: [error 164] LEGION ERROR: Dynamic type mismatch in 'get_index_space_domain' (from file /Users/mpapadakis/legate.core/legion/runtime/legion/region_tree.inl:2843)
```

This might require changes in the core.

**Comment by ipdemes** [comment]
@manopapad I have reproduced the error and it seems it might take some time to fix it. Should we create a separate GitHub issue and PR for this use case?

**Comment by manopapad** [comment]
Yes, let me update the testcase to skip the problematic cases, and I'll open a new issue about this.

**Comment by ipdemes** [comment]
@manopapad : do you think we can merge it when CI pass?

**Comment by manopapad** [comment]
@ipdemes I rebased the PR onto branch-22.10, and reformatted some testcases following guidance from @bryevdv. Let's verify that these pass before merging. Note that you will now need the new `pytest-lazy-fixture` package to run these tests.
